### ISSUE TITLE
1.1.395: `TypeForm` does not accept certain `UnionType`s

### ISSUE THREAD

**Issue by JacobHayes** [microsoft/pyright#9985]
**Describe the bug**

A function hinted to accept a `TypeForm[Any]` works when passed a direct union type expressions, but does not seem to accept "derived" expressions created from a parameter of `type[T]`. Trying to follow along with https://github.com/microsoft/pyright/discussions/9397, I _think_ this should be accepted and it was allowed in 1.1.394. I know `TypeForm` is experimental, just trying to get a feel for working with the PEP.

**Code**

```python
from types import UnionType
from typing import TYPE_CHECKING, Any, get_origin

if TYPE_CHECKING:
    from typing_extensions import TypeForm


def is_union_type(t: "TypeForm[Any]"):
    print(get_origin(t) is UnionType)


t = int
is_union_type(t)  # False
is_union_type(t | None)  # True


def check[T](t: type[T]):
    is_union_type(t)  # t=int: False
    is_union_type(t | None)  # t=int: True


check(int)
```

Changing `is_union_type` to `is_union_type(t: "TypeForm[Any] | UnionType"):` fixes things, but I wouldn't expect that to be required when `is_union_type(t | None)` works without it.

**VS Code extension or command-line**

Works with pyright 1.1.394 (cli), but errors with 1.1.395 (lsp).

**Comment by erictraut** [comment]
Pyright is working correctly here and is complying with the PEP (of which I'm a coauthor). 

You can't use variables in type expressions, so `t | None` is not valid for `TypeForm`. The expression `is_union_type(type[T] | None)` is valid because `type[T] | None` is a valid type expression.

**Comment by JacobHayes** [comment]
Makes sense! I've seen the `Variable not allowed in type expression` error before - should that be raised here if that's the root cause?

Also, does that mean the `def isassignable` example from the PEP cannot be used in any function that takes a `type[...]` parameter? I was expecting to be able to write `isinstance`/`issubclass` like functions (which can be called anywhere, indeed `issubclass(int, t | None)` works fine inside `check`), but if they can only operate over exact type expressions, that's a pretty big (but understandable) limitation - or is the suggestion to use `TypeForm[Any] | UnionType` in the type hint?

**Comment by erictraut** [comment]
If you've read #9397, you know that I was against adding the special-case rule that `type[T]` is a subtype of `TypeForm[T]`, but I bowed to the majority consensus. This special case introduces an unsoundness into the TypeForm mechanism. Part of my argument here is that we decided _not_ to support special cases for other classes that are commonly used to encode type forms. That includes `UnionType`. The reason we didn't want to support these other types is that they can encode illegal type forms. (For example: `Union[1, 2]`, which does not describe a legal type in the Python type system but is an instance of UnionType.).

If you want your function to accept a `UnionType` and are willing to live with this unsoundness or add all of the necessary logic to validate that the UnionType encodes a valid type, you can indicate that your function accepts `TypeForm | UnionType`.
